go/vt/discovery/endpointwatcher.go

```diff
@@ -0,0 +1,104 @@
+package discovery
+
+import (
+	"time"
+
+	log "github.com/golang/glog"
+	pbt "github.com/youtube/vitess/go/vt/proto/topodata"
+	"github.com/youtube/vitess/go/vt/topo"
+	"golang.org/x/net/context"
+)
+
+// NewEndPointWatcher returns an EndPointWatcher and starts refreshing.
+func NewEndPointWatcher(topoServer topo.Server, hc HealthCheck, cellsToWatch []string, refreshInterval time.Duration) *EndPointWatcher {
+	epw := &EndPointWatcher{
+		topoServer:      topoServer,
+		hc:              hc,
+		cellsToWatch:    cellsToWatch,
+		refreshInterval: refreshInterval,
+		endPoints:       make(map[string]*pbt.EndPoint),
+	}
+	epw.start()
+	return epw
+}
+
+// EndPointWatcher pulls endpoints of all running tablets periodically.
+type EndPointWatcher struct {
+	topoServer      topo.Server
+	hc              HealthCheck
+	cellsToWatch    []string
+	refreshInterval time.Duration
+	cancelFunc      context.CancelFunc
+	endPoints       map[string]*pbt.EndPoint
+}
+
+// start starts the pulling.
+func (epw *EndPointWatcher) start() {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	epw.cancelFunc = cancelFunc
+	go epw.watch(ctx)
+}
+
+// watch pulls all endpoints and notifies HealthCheck by adding/removing endpoints.
+func (epw *EndPointWatcher) watch(ctx context.Context) {
+	ticker := time.NewTicker(epw.refreshInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			watchEndPoints := make(map[string]*pbt.EndPoint)
+			for _, cell := range epw.cellsToWatch {
+				tabletAlias, err := epw.topoServer.GetTabletsByCell(ctx, cell)
+				if err != nil {
+					select {
+					case <-ctx.Done():
+						return
+					default:
+					}
+					log.Errorf("cannot get tablets for cell: %v: %v", cell, err)
+					continue
+				}
+				for _, alias := range tabletAlias {
+					tablet, err := epw.topoServer.GetTablet(ctx, alias)
+					if err != nil {
+						select {
+						case <-ctx.Done():
+							return
+						default:
+						}
+						log.Errorf("cannot get tablets for cell %v: %v", cell, err)
+						continue
+					}
+					endPoint, err := topo.TabletEndPoint(tablet.Tablet)
+					if err != nil {
+						log.Errorf("cannot get endpoint from tablet %v: %v", tablet, err)
+					}
+					key := endPointToMapKey(endPoint)
+					if _, ok := epw.endPoints[key]; !ok {
+						// new endpoint
+						epw.endPoints[key] = endPoint
+						epw.hc.AddEndPoint(cell, endPoint)
+					}
+					watchEndPoints[key] = endPoint
```

Comment:
this is very confusing, how about this:
- epw.endPoints has the 'old map'
- rename watchEndPoints to newEndPoints, put the new one in there, and remove them from the old one as you do
- go throw the old map and remove everything that's left
- save the new map into epw.endPoints.

Otherwise the operations in the map are impossible to understand, the way they are now...